doc/subsystems/codegen/codegen.rst

```diff
@@ -0,0 +1,409 @@
+
+.. function:: codegen.device_tree_controller_compatible_x(device_type_name, device_index, compatible [, default="<unset>"])
+
+    Check compatibility to device given by type and index.
+
+    :param device_type_name: Type of device controller (e.g. 'SPI', 'GPIO', 'PIN', ...)
+    :param device_index: Index of device
+    :param compatible: driver compatibility string (e.g. st,stm32-spi-fifo)
+    :return: 1 if compatible, 0 otherwise
+
+.. function:: codegen.device_tree_controller_compatible(device_type_name, compatible)
+
+    Check compatibility to at least one activated device of given type.
+
+    The compatible parameter is checked against the compatible property of
+    all activated devices of given type.
+
+    :param device_type_name: Type of device controller (e.g. 'SPI', 'GPIO', 'PIN', ...)
+    :param compatible: driver compatibility string (e.g. st,stm32-spi-fifo)
+    :return: 1 if there is compatibility to at least one activated device,
+             0 otherwise
+
+.. function:: codegen.device_tree_controller_data_name(device_type_name, device_index, data)
+
+    Get the name of the driver data.
+
+    Generates an unique name for driver data.
+
+    :param device_type_name: Type of device controller (e.g. 'SPI', 'GPIO', 'PIN', ...)
+    :param device_index: Index of device
+    :param data: suffix for data (e.g. 'config')
+    :return: controller data name (e.g. ST_STM32_SPI_FIFO_4000000_config)
+
+.. function:: codegen.device_tree_controller_device_name(device_type_name, device_index)
+
+    Get the device name.
+
+    The device tree prefix of the device is used as the device name.
+
+    :param device_type_name: Type of device controller (e.g. 'SPI', 'GPIO', 'PIN', ...)
+    :param device_index: Index of device
+    :return: device name (e.g. ST_STM32_SPI_FIFO_4000000)
+
+.. function:: codegen.device_tree_controller_driver_name(device_type_name, device_index)
+
+    Get the driver name.
+
+    This is a convenience function for:
+
+    - codegen.device_tree_controller_property(device_type_name, device_index, 'LABEL')
+
+    :param device_type_name: Type of device controller (e.g. 'SPI', 'GPIO', 'PIN', ...)
+    :param device_index: Index of device
+    :return: driver name (e.g. "SPI_0")
+
+.. function:: codegen.device_tree_controller_driver_name_indirect(device_type_name, device_index, property_name)
+
+     Get the driver name of another device given by the property.
+
+    :param device_type_name: Type of device controller (e.g. 'SPI', 'GPIO', 'PIN', ...)
+    :param device_index: Index of device
+    :param property_name: Property that denotes the device tree prefix of the other device
+    :return: driver name of other device (e.g. "GPIOA")
+
+Guarding chunks of source code
+------------------------------
+
+.. function:: codegen.if_device_tree_controller_compatible(device_type_name, compatible)
+
+    Stop code generation if there is no activated device that is compatible.
```

Comment:
What does "stop" mean here?  Just that no output for that chunk is generated or the whole code generation processing stops?

Reply:
Whole code processing stops